NEW: refactor for sample peds for implementation of Feature peds

cherman2 opened the pull request:
This is a refactor of sample PEDS for new implementation of feature peds. This PR also includes the implementation of Feature peds

Feature peds tracks what percentage of samples have a specific feature from the donor to investigate what features are engrafting in recipients.

cherman2 committed:
intial filtering refactor

cherman2 committed:
sample peds refactor for feature peds

cherman2 commented:
This pr currently has the sample peds refactor implemented and tested. As well as draft feature peds code with no unit test nor hookup to the q2-interface

cherman2 committed:
Fix issue with using decommissioned method for tests

cherman2 committed:
add feature-peds hookup and add metadata.filter

cherman2 commented:
Need to add Unit tests for feature-ped and ~~solve issue with taxonomic label being cut off~~ (This will be a new PR)

cherman2 committed:
adds usage example for feature peds

cherman2 committed:
refactor tests and add feature-peds unit test

lizgehret commented:
> Need to add Unit tests for feature-ped and solve issue with taxonomic label being cut off

@cherman2 I see that you've since added tests for feature-peds, but does the taxonomic label being cut off need to be addressed in this PR? Or are you going to open up a separate PR for that?

cherman2 commented:
@lizgehret, I talked with evan and we decided it made more sense as a separate PR

lizgehret commented on q2_fmt/_peds.py:
```diff
@@ -8,30 +8,191 @@
+                                    filter_missing_references,
+                                    reference_column)
+    subject_series = _check_subject_column(metadata, subject_column)
+    _check_dupilicate_subject_timepoint(subject_series, metadata,
```
```suggestion
    _check_duplicate_subject_timepoint(subject_series, metadata,
```

lizgehret commented on q2_fmt/_peds.py:
```diff
@@ -8,30 +8,191 @@
+                                    filter_missing_references,
+                                    reference_column)
+    peds_df = pd.DataFrame(columns=['id', 'measure', 'recipients with feature',
+                                    'all possible recipients with feature',
```
What's the difference between 'recipients with feature' and 'all possible recipients with feature'? Is this related to the recipients' donor?

cherman2 replied:
Recipient with features is the number of recipients that ended up with the donor feature, while all possible recipients with feature is the amount of people who could have received the feature

lizgehret commented on q2_fmt/_peds.py:
```diff
@@ -8,30 +8,191 @@
+                                        subject_column, time_column)
+    # return things that should be removed
+    metadata, used_references = \
+        _check_subjects_in_all_timepoint(subject_series, num_timepoints,
```
Should this method be `_check_subjects_in_all_timepoints` with timepoint as plural?

cherman2 replied:
I changed this not through github because I wanted to change all the calls of it.

lizgehret commented on q2_fmt/_peds.py:
```diff
@@ -46,12 +207,20 @@ def sample_peds(table: pd.DataFrame, metadata: qiime2.Metadata,
+    return subject_series
+
+
+def _check_dupilicate_subject_timepoint(subject_series, metadata,
```
```suggestion
def _check_duplicate_subject_timepoint(subject_series, metadata,
```

lizgehret commented on q2_fmt/_peds.py:
```diff
@@ -8,30 +8,191 @@
+                  subject_column: str,
+                  reference_column: str) -> (pd.DataFrame):
+    table = table > 0
+    donordf = table[table.index.isin(reference_series)]
```
Not sure if the juice is worth the squeeze, but it might be nice to keep all of your variable naming conventions the same (i.e. snake case between words vs. no word separation like: donor_df vs. donordf)

cherman2 replied:
Did this in commit: [caf9a30](https://github.com/qiime2/q2-fmt/pull/48/commits/caf9a301c8fc7445945ae5177236c4bf478996bb)

lizgehret commented on q2_fmt/_peds.py:
```diff
@@ -8,30 +8,191 @@
+        })
+        peds_df['measure'].attrs.update({
+            'title': "PEDS",
+            'description': 'Proportional Engraftment of Donor Strains '
```
```suggestion
            'description': 'Proportional Engraftment of Donor Strains'
```

lizgehret commented on q2_fmt/_peds.py:
```diff
@@ -8,30 +8,191 @@
+        })
+        peds_df["subject"].attrs.update({
+            'title': subject_column,
+            'description': 'ID to link samples across time'
```
```suggestion
            'description': 'Subject IDs linking samples across time'
```
This seemed more clear to me, but feel free to take it or leave it 🤷‍♀️

lizgehret commented on q2_fmt/_peds.py:
```diff
@@ -8,30 +8,191 @@
+                                         donor_sum[count], peds_time, feature]
+            peds_df = peds_df.dropna()
+        peds_df['id'].attrs.update({
+            'title': "FeatureID",
```
```suggestion
            'title': "Feature ID",
```

lizgehret commented on q2_fmt/_peds.py:
```diff
@@ -8,30 +8,191 @@
+        })
+        peds_df['measure'].attrs.update({
+            'title': "PEDS",
+            'description': 'Proportional Engraftment of Donor Strains '
```
```suggestion
            'description': 'Proportional Engraftment of Donor Strains'
```

lizgehret commented on q2_fmt/_peds.py:
```diff
@@ -8,30 +8,191 @@
+            'description': 'Time'
+        })
+        peds_df['subject'].attrs.update({
+            'title': "FeatureID",
```
```suggestion
            'title': "Feature ID",
```

lizgehret commented on q2_fmt/_peds.py:
```diff
@@ -8,30 +8,191 @@
+    return num_timepoints
+
+
+def _check_time_column_numeric(metadata, time_column):
```
I'm interested in what @ebolyen thinks about this - but I'm wondering if you can just combine the two filtering methods for the time column - checking for existence, and then the type (ensuring numeric) before returning it from the `_check_for_time_column` helper method.

cherman2 replied:
I am open to merging them. I just kinda like that it was only checking one specific thing. Made more sense in my head

lizgehret commented on q2_fmt/_peds.py:
```diff
@@ -8,30 +8,191 @@
+    return reference_series
+
+
+def _check_associated_reference(reference_series, metadata, time_column,
```
```suggestion
def _filter_associated_references(reference_series, metadata, time_column,
```
Also can take it or leave it on this suggestion, but it might be a bit more clear to call this one a `filter` helper instead of a `check` helper - since you aren't checking for something specific here, just pulling out a subset of `reference_series` and returning that with the associated metadata.

cherman2 replied:
Changed this outside of github so I could address when it was called too.

lizgehret commented on q2_fmt/_peds.py:
```diff
@@ -61,15 +230,20 @@ def sample_peds(table: pd.DataFrame, metadata: qiime2.Metadata,
                              ' timepoints: %s' % (subject, timepoint_list))
 
+
+def _check_subjects_in_all_timepoint(subject_series, num_timepoints,
```
```suggestion
def _check_subjects_in_all_timepoints(subject_series, num_timepoints,
```

lizgehret commented on q2_fmt/_peds.py:
```diff
@@ -79,16 +253,77 @@ def sample_peds(table: pd.DataFrame, metadata: qiime2.Metadata,
-    return peds_df
 
+# PEDS calculation methods
+def _create_recipient_table(refernce_series, time_metadata, tabledf):
```
```suggestion
def _create_recipient_table(reference_series, time_metadata, tabledf):
```

lizgehret commented on q2_fmt/_peds.py:
```diff
@@ -79,16 +253,77 @@ def sample_peds(table: pd.DataFrame, metadata: qiime2.Metadata,
+# PEDS calculation methods
+def _create_recipient_table(refernce_series, time_metadata, tabledf):
+    subset_reference_series = \
+        refernce_series[refernce_series.index.isin(time_metadata.index)]
```
```suggestion
        reference_series[reference_series.index.isin(time_metadata.index)]
```

lizgehret commented on q2_fmt/_visualizer.py:
```diff
@@ -17,6 +17,7 @@ def plot_heatmap(output_dir: str, data: pd.DataFrame):
     J_ENV = jinja2.Environment(
         loader=jinja2.PackageLoader('q2_fmt', 'assets')
     )
+    print(data)
```
```suggestion
```

lizgehret commented on q2_fmt/tests/test_engraftment.py:
```diff
@@ -873,3 +858,54 @@ def test_unique_subjects_in_timepoints(self):
+                                       time_column="group",
+                                       reference_column="Ref",
+                                       subject_column="subject")
+        print(feature_peds_df)
```
```suggestion
```

lizgehret reviewed (commented):
Overall this seems reasonable! Some small comments inline, nothing major.

cherman2 committed:
Update q2_fmt/_peds.py

Co-authored-by: Liz Gehret <54517601+lizgehret@users.noreply.github.com>

cherman2 committed:
Update q2_fmt/_peds.py

Co-authored-by: Liz Gehret <54517601+lizgehret@users.noreply.github.com>

cherman2 committed:
Update q2_fmt/_peds.py

Co-authored-by: Liz Gehret <54517601+lizgehret@users.noreply.github.com>

cherman2 committed:
Update q2_fmt/_peds.py

Co-authored-by: Liz Gehret <54517601+lizgehret@users.noreply.github.com>

cherman2 committed:
Update q2_fmt/_peds.py

Co-authored-by: Liz Gehret <54517601+lizgehret@users.noreply.github.com>

cherman2 committed:
Update q2_fmt/tests/test_engraftment.py

Co-authored-by: Liz Gehret <54517601+lizgehret@users.noreply.github.com>

cherman2 committed:
Update q2_fmt/_peds.py

Co-authored-by: Liz Gehret <54517601+lizgehret@users.noreply.github.com>

cherman2 committed:
Update q2_fmt/_peds.py

Co-authored-by: Liz Gehret <54517601+lizgehret@users.noreply.github.com>

cherman2 committed:
Update q2_fmt/_peds.py

Co-authored-by: Liz Gehret <54517601+lizgehret@users.noreply.github.com>

cherman2 committed:
Update q2_fmt/_visualizer.py

Co-authored-by: Liz Gehret <54517601+lizgehret@users.noreply.github.com>

cherman2 committed:
Update q2_fmt/_peds.py

Co-authored-by: Liz Gehret <54517601+lizgehret@users.noreply.github.com>

cherman2 committed:
Update q2_fmt/_peds.py

Co-authored-by: Liz Gehret <54517601+lizgehret@users.noreply.github.com>

cherman2 committed:
minor review suggestions

cherman2 committed:
variable naming consistency and importing wrong function in unit test

cherman2 committed:
stupid

cherman2 committed:
another one

cherman2 commented:
@lizgehret, This is ready for second review

lizgehret commented:
Okay, technical review lgtm @cherman2 - I'm going to pull this down locally to test and then this should be g2g!

lizgehret commented:
Help text looks good, general file structure for output QZA looks good. Some questions/comments post-testing:

- Should any column be allowed under `--p-subject-column`?
- When using a numeric column under `--p-reference-column` (e.g. time_point is what I used, testing with your example data) the following error occurs:
![Screen Shot 2023-05-01 at 3 03 49 PM](https://user-images.githubusercontent.com/54517601/235539564-9d9114fa-1832-4458-8852-05b6f0680f25.png)
I tested this with `time_point` and `SubjectID`, which both produced the same error. When I tried using `Sample ID`, I got this (slightly different) error:
![Screen Shot 2023-05-01 at 3 07 43 PM](https://user-images.githubusercontent.com/54517601/235540013-c14575f6-7c74-4af1-ba94-028b5ca6227f.png)
Not sure what limitations are intended for this column, but it might be worth adding a guard that will raise an error if you don't want any numeric columns being used here. Additionally, I might add a more verbose description in the latter error message for what should be used as a reference column.
- I tested out using a non-numeric column in `--p-time-column`, which produced the following error message:
```
Donor must be numeric
```
I would include the column name in this error message, along with the actual column that was included in the command (just for clarity's sake). Maybe something like:
```
Non-numeric values found in `--p-time-column`. Please make sure the column selected contains only simple integer values. Column with non-numeric values that was selected: "Donor"
```

lizgehret reviewed (changes requested):
Some comments above post-user review! Just FYI I used the usage example dataset to test - it might be worth using a more 'real-world' dataset to test on after this round just to see if the actual results (vs. just the general file structure, etc) make sense.

cherman2 committed:
addressing user review comments

cherman2 commented:
Thanks @lizgehret for your review! 
This ended up being a lot more changes then I was expecting so I thought I would list out what I changed to address your comments. 

1. I added a check subject column name function to feature peds. It already existed but I wasn't using it in feature peds. 
2. I wrote, and added to FP and SP, a new function that checks that subject column is not numeric.
        a.  I fixed my usage example metadata because the subject values were numeric lol.
        b. I added a unit-test for testing if subject column is numeric.
3. I wrote, and added to FP and SP, a new function that checks that reference column is not numeric. 
        a. I added a unit-test for testing if reference column is numeric.
4. I refactored the function that tests if time column is numeric to match how I checked metadata types in the above new functions.
         a. I added a unit-test for testing if time column is not numeric
5. I added a check that makes sure that the references are in the feature table before making my donor dataframe. This addresses the vague error you got when you used time_point as the reference. 
          a. I added a test to make sure this catches and lists reference values that are not in the feature table.
6. I now check to see if time, reference or subject are in the index before I say that the column name does not exist. This address the error you got when you put Sample ID as your reference name. 
          a.  I added 3 new unit test to check that time column, reference column, and subject column are not the metadata                               index
7. I re-wrote the "`blank` wasn't found in the metadata" error so that it says what parameter was being checked. (e.g. instead of saying "there was an error finding Sample ID in the metadata". It now says "There was an error finding the provided `--p-reference-column`: `Sample ID` in the metadata'".
      a. I rewrote the  test_incorrect_group_column_name unit-test for this new error message 
      b. I rewrote the test_incorrect_reference_column_name unit-test for this new error message
      c. I rewrote the test_incorrect_subject_column_name unit-test for this new error message
8. I implemented your suggested rewrite to my time column is not numeric error.

cherman2 committed:
minor fix to error message

lizgehret commented on q2_fmt/_peds.py:
```diff
@@ -170,24 +186,43 @@ def _check_for_time_column(metadata, time_column):
 
 
-def _check_time_column_numeric(metadata, time_column):
+def _check_time_column_numeric(column_properties, time_column):
```
Hey @cherman2, I think you re-use iterations of this helper method enough times that it would be worth a slight refactor (post-2023.5, since I know we're right up against a deadline on this). I'd re-write something like this:
```
def _check_column_type(column_properties, column, column_type):
    try:
        assert column_properties[column].type == column_type
    except AssertionError as e:
        raise AssertionError('Non-%s values found in `--p-%s-column`. Please make sure the column selected contains the correct MetadataColumn type. Column with non-%s values that was selected: %s' % (column_type, column, column_type, column)) from e
```
I'm not sure if the syntax of the AssertionError message is exactly correct, but this is mainly just to show an example for how I'd refactor all of these helpers.

lizgehret commented on q2_fmt/_peds.py:
```diff
@@ -170,24 +186,43 @@ def _check_for_time_column(metadata, time_column):
     except Exception as e:
-        raise KeyError('There was an error finding %s in the metadata'
-                       % reference_column) from e
+        if reference_column == metadata.index.name:
```
You could also refactor this branch and the `subject_column` one to be another helper that checks if the given column is equal to the `metadata.index.name`.

lizgehret commented on q2_fmt/_peds.py:
```diff
@@ -256,6 +308,17 @@ def _check_subjects_in_all_timepoints(subject_series, num_timepoints,
 
+def _check_subject_column_categorical(column_properties, subject_column):
+    try:
+        print(column_properties[subject_column].type)
```
```suggestion
```

lizgehret commented on q2_fmt/tests/test_engraftment.py:
```diff
@@ -909,3 +915,125 @@ def test_sample_id_match(self):
                          reference_column="Ref",
                          subject_column="subject")
+
+    def test_group_column_categorical(self):
```
With the helper method refactor mentioned above, you won't need all of these individual column/type tests either - you can just have one that confirms the behavior given a specific column and column type.

lizgehret commented on q2_fmt/tests/test_engraftment.py:
```diff
@@ -909,3 +915,125 @@ def test_sample_id_match(self):
+                          time_column="group", reference_column="Ref",
+                          subject_column="subject")
+
+    def test_subject_column_name_is_ID(self):
```
Similar to last comment, if you refactor the subject/reference column check in metadata to be a single helper method, you can pair this down to just one test instead of two.

lizgehret reviewed (approved):
Hey @cherman2 overall this looks good to merge! A few suggestions in-line for some small refactors post-2023.5 release, since I realize this is time sensitive and none of the suggestions impact your code's actual functionality.

Only thing that you can actually update before I merge is a small print statement that got left behind in your debugging process. I included that as a suggestion in my inline comments, so you can just commit that and this will be g2g. Thanks for all of your hard work on this!

cherman2 committed:
Update q2_fmt/_peds.py

Co-authored-by: Liz Gehret <54517601+lizgehret@users.noreply.github.com>
